### DISCUSSION TITLE
Shopify API PHP Library

### DISCUSSION THREAD

**Discussion by jsirish** [dynamic/silverstripe-shopify#6] [Ideas]
Looking for thoughts on the API library choice:

1. PHP library to use
2. Where Shopify API credentials should be set

@muskie9 @mak001 

---

The module will include a PHP API client to interact with the [Shopify Admin API](https://shopify.dev/docs/admin-api). The Admin API provides extensive access to data about individual Shopify stores, and will allow us to add custom features to the Shopify user experience. 

- Read and write Shopify store information, including data about products, inventory, orders, shipping, and more.
- The Admin API is available through GraphQL and REST.

The ideal solution would be to serve as a wrapper for a mature, maintained Shopify PHP API library. Shopify lists existing [Shopify PHP API classes](https://shopify.dev/tools/libraries/admin-api#php) that can be used to easily interface with the Admin API.

Initial proposed library to explore is [phpclassic/php-shopify](https://github.com/phpclassic/php-shopify). This library appears to be mature, active, and was referenced most often during initial Shopify PHP research in articles and message boards. This module seems to offer both REST and GraphQL interfaces, Dynamic will need to decide which route to pursue.

API credentials would ideally be set per server in the environment file. If this is impractical, credentials could be stored in config yaml files, with different credentials set per website environment mode - typically one set for Test, one set for Live.
The client would utilize functions from the PHP library to sync data between the website and Shopify, such as product records and customer information. 

The client would utilize functions from the PHP library to sync data between the website and Shopify, such as product records and customer information.

**Comment by mak001** [comment]
Just to write down thoughts on rest vs graphql: 

Rest will return all the information, regardless of if the information is needed. This would make implementing new features that require fields outside of what is used easier than graphql.

GraphQL will only return fields that are requested, so there may be more we have to do on our end to make it configurable to allow more fields to be queried. It will also take a little bit more work to implement features that require more fields to be queried, but i think it would be slightly better because it would make us think of what fields are required for each feature.

**Comment by jsirish** [comment]
Also, it'd be great to come up with an initial list of API client functions. The initial models to sync are likely Product, Variant, Collection and Customer.

**Comment by mak001** [comment]
More on the graphql vs rest:
We will need to take into consideration the rate limits for each.  https://shopify.dev/concepts/about-apis/rate-limits
rest would require more queries to shopify than graphql, but graphql queries require more work. For example this is a graphql query to get orders of a customer, with some of the data:
```graphql
{
  customers(first: 1, query: "email:bob.slydell@example.com") {
    edges {
      node {
        id
        displayName
        orders(first: 10) {
          edges {
            node {
              taxLines {
                rate
                ratePercentage
                title
              }
              lineItems(first: 10) {
                edges {
                  node {
                    sku
                    quantity
                    taxLines {
                      rate
                      ratePercentage
                      title
                    }
                    totalDiscountSet {
                      presentmentMoney {
                        amount
                        currencyCode
                      }
                    }
                    originalTotalSet {
                      presentmentMoney {
                        amount
                        currencyCode
                      }
                    }
                  }
                }
              }
            }
          }
        }
      }
    }
  }
}

```


Here is the response:

```json
{
  "data": {
    "customers": {
      "edges": [
        {
          "node": {
            "id": "gid://shopify/Customer/1310260264982",
            "displayName": "Bob Slydell",
            "orders": {
              "edges": [
                {
                  "node": {
                    "taxLines": [],
                    "lineItems": {
                      "edges": [
                        {
                          "node": {
                            "sku": "",
                            "quantity": 1,
                            "taxLines": [],
                            "totalDiscountSet": {
                              "presentmentMoney": {
                                "amount": "0.0",
                                "currencyCode": "CAD"
                              }
                            },
                            "originalTotalSet": {
                              "presentmentMoney": {
                                "amount": "6359.0",
                                "currencyCode": "CAD"
                              }
                            }
                          }
                        },
                        {
                          "node": {
                            "sku": "",
                            "quantity": 1,
                            "taxLines": [],
                            "totalDiscountSet": {
                              "presentmentMoney": {
                                "amount": "0.0",
                                "currencyCode": "CAD"
                              }
                            },
                            "originalTotalSet": {
                              "presentmentMoney": {
                                "amount": "7043.0",
                                "currencyCode": "CAD"
                              }
                            }
                          }
                        }
                      ]
                    }
                  }
                }
              ]
            }
          }
        }
      ]
    }
  },
  "extensions": {
    "cost": {
      "requestedQueryCost": 445,
      "actualQueryCost": 17,
      "throttleStatus": {
        "maximumAvailable": 1000,
        "currentlyAvailable": 952,
        "restoreRate": 50
      }
    }
  }
}
```

**Reply by jsirish** [reply]
My only concern with going with REST would be if it gets depreciated in favor of GraphQL (Storefront API only offers GraphQL) but I haven't seen anything indicating that's on the roadmap. So I think REST would work

**Reply by muskie9** [reply]
There was mention on our call the week of 3/8 with Shopify that the rate limit could be increased, as well as other ways to fetch data in efficient ways (leveraging their infrastructure more) which may make GQL worth the time now. If we can identify the immediate needs from a query perspective as it relates to the data needed for various fetches, we could work with Shopify to get the initial build across the line and fine tune or expand the query offering later.

I think this will be less flexible at the start, however knowing they will be pushing for GQL more in the long run it's probably better to put our time there vs REST and later transitioning.

**Comment by muskie9** [comment]
I did a bit of poking around in packagist and did find 2 other active packages for shopify that appear to have had commits/releases in 2021:

- https://github.com/osiset/Basic-Shopify-API
- https://github.com/slince/shopify-api-php

They have a smaller shown "use" and "contributor" base, however I think we're good to try the first mentioned in this discussion and if we find we need to try something else these are options. We should keep an eye on activity of the phpclassic package so we see how active it is in relation to the others to determine if the activity of the others is shopify related, or patch related for those additional packages.

**Reply by muskie9** [reply]
After a short period looking into the phpclassic package, it seems only the `POST` method is allowed. My guess is this is for this module only at this point so I'll take a look at the osiset module for a bit to see if that supports additional methods for GQL.

**Reply by muskie9** [reply]
With a little playing around I have the osiset package working with GQL. I'll keep exploring this route to see if it will be viable as the package of choice.

**Comment by jsirish** [comment]
Works for me!

**Comment by muskie9** [comment]
Circling back around to this, there is a PHP library provided by Shopify. I don't believe I saw it in reviewing this, however it may be worth looking into for a future major version.

- https://packagist.org/packages/shopify/shopify-api
- https://shopify.dev/api/admin-graphql#client_libraries

I didn't get a chance to install and try anything with it as it requires gulp 7 and our current shopify module requires gulp 6. This may be something to dig into as a part of the sync task re-write.
